ddsrecorder_participants/src/cpp/mcap/McapHandler.cpp

```diff
@@ -0,0 +1,460 @@
+    }
+
+    try
+    {
+        auto channel_id = get_channel_id_(topic);
+        msg.channelId = channel_id;
+    }
+    catch(const utils::Exception& e)
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Schema for topic " << topic << " not yet available, inserting to pending samples queue.");
+
+        std::lock_guard<PendingSamplesMapType> lock(pending_samples_);
+
+        if (pending_samples_[topic.type_name].size() == max_pending_samples_)
+        {
+            pending_samples_[topic.type_name].pop();
+        }
+        pending_samples_[topic.type_name].push({topic.m_topic_name, msg});
+
+        throw;
+    }
+
+    try
+    {
+        add_data_(msg);
+    }
+    catch(const utils::Exception& e)
+    {
+        throw utils::InconsistencyException(
+            STR_ENTRY << "Error writting in MCAP a message in topic " << topic.m_topic_name
+        );
+    }
+}
+
+void McapHandler::start()
+{
+    if (!paused_)
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Ignoring start command, instance already started.");
+    }
+    else
+    {
+        logInfo(
+            DDSRECORDER_MCAP_HANDLER,
+            "Starting handler.");
+
+        paused_.store(false);
+        stop_event_thread_();
+    }
+}
+
+void McapHandler::pause()
+{
+    if (paused_)
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Ignoring pause command, instance already paused.");
+    }
+    else
+    {
+        logInfo(
+            DDSRECORDER_MCAP_HANDLER,
+            "Pausing handler.");
+
+        dump_data_();
+
+        paused_.store(true);
+        event_thread_ = std::thread(&McapHandler::event_thread_routine_, this);
+    }
+}
+
+void McapHandler::trigger_event()
+{
+    if (!paused_)
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Ignoring trigger event command, instance is not paused.");
+    }
+    else
+    {
+        logInfo(
+            DDSRECORDER_MCAP_HANDLER,
+            "Triggering event.");
+        {
+            std::lock_guard<std::mutex> lock(event_cv_mutex_);
+            event_triggered_ = true;
+        }
+        event_cv_.notify_one();
+    }
+}
+
+void McapHandler::add_data_(
+        const Message& msg)
+{
+    std::lock_guard<std::mutex> guard(write_mtx_);
+
+    samples_buffer_.push(msg);
+    if (!paused_ && samples_buffer_.size() == buffer_size_)
+    {
+        logInfo(DDSRECORDER_MCAP_HANDLER, "Full buffer, writting to disk...");
+        dump_data_nts_();
+    }
+}
+
+void McapHandler::add_pending_samples_(
+        const std::string& schema_name)
+{
+    assert(pending_samples_.find(schema_name) != pending_samples_.end());
+    auto& pending_queue = pending_samples_[schema_name];
+
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Sending pending samples of type: " << schema_name << ".");
+
+    while (!pending_queue.empty())
+    {
+        auto& sample = pending_queue.front();
+        DdsTopic sample_topic;
```

Comment:
Can't this be created outside the loop?

Reply:
Same for channel id

Reply:
The idea was related to not set the topic and channel id each time, cause I thought all these messages belong to the same channel